### PR TITLE
feat(event_handler): support to enable or disable compression in custom responses

### LINKED ISSUE #2513: Feature request: APIGatewayRestResolver exception_handler compress data
### Expected Behaviour

exception_handler does not compress response

### Current Behaviour

exception_handler compress response if decorator contained `compress=True`

### Code snippet

```python
from aws_lambda_powertools.event_handler import content_types

@app.exception_handler(Exception)
def handle_invalid_params(ex: Exception):
    response = Response(
            status_code=500,
            content_type=content_types.APPLICATION_JSON,
            body=json.dumps({"ex": str(ex)})
        )
    return response 

@app.get("/", compress=True)
def foo(datasetid: str):
    raise Exception
```

### Possible Solution

_No response_

### Steps to Reproduce

Run code snippet

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.7

### Packaging format used

Serverless Application Repository (SAR) App

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #2513 

### Changes

Use case: Customers use the `exception_handler` decorator to handle 4xx/5xx errors. They might want to encode non-2xx responses.

This pull request improve the `Response` class with a new `compress` parameter to override compression setting defined in the route.

### User experience

**BEFORE**

```python
import json

from aws_lambda_powertools.event_handler import (
    APIGatewayRestResolver,
    Response,
    content_types,
)
from aws_lambda_powertools.utilities.typing import LambdaContext

app = APIGatewayRestResolver()

@app.exception_handler(NameError)
def handle_invalid_params(ex: NameError):
    response = Response(
            status_code=500,
            content_type=content_types.APPLICATION_JSON,
            body=json.dumps({"error": "error"})
        )
    return response 

@app.get("/hello", compress=True)
def hello():
    raise NameError("Error")


def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

**AFTER**
```python
import json

from aws_lambda_powertools.event_handler import (
    APIGatewayRestResolver,
    Response,
    content_types,
)
from aws_lambda_powertools.utilities.typing import LambdaContext

app = APIGatewayRestResolver()

@app.exception_handler(NameError)
def handle_invalid_params(ex: NameError):
    response = Response(
            status_code=500,
            content_type=content_types.APPLICATION_JSON,
            body=json.dumps({"error": "error"}),
            compress=False
        )
    return response 

@app.get("/hello", compress=True)
def hello():
    raise NameError("Error")


def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
looking

**Comment by heitorlessa** [comment]
quick comment: Update the example **not** to send stacktrace to prevent **data leak**; log the stack trace instead, and return another message instead back to the client.

File: aws_lambda_powertools/event_handler/api_gateway.py [hunk 1]
```diff
@@ -250,12 +252,33 @@ def _route(self, event: BaseProxyEvent, cors: Optional[CORSConfig]):
             self._add_cors(event, cors or CORSConfig())
         if self.route.cache_control:
             self._add_cache_control(self.route.cache_control)
-        if self.route.compress and "gzip" in (event.get_header_value("accept-encoding", "") or ""):
+        if (
+            self.route.compress
+            and "gzip" in (event.get_header_value("accept-encoding", "") or "")
+            and self.response.compress is not False
+        ):
+            self._compress()
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py]: could you break this down to reduce cognitive overload?

A private method to control whether to enable compression will be easier to reason. 

e.g., 

`if self._has_compression_enabled(...): self._compress()`

**Review by heitorlessa — CHANGES_REQUESTED** [review]
This is great!! Only one major change to reduce cognitive overload and make it easier to extend new functionalities later.

File: tests/functional/event_handler/test_api_gateway.py [hunk 2]
```diff
@@ -366,6 +366,58 @@ def test_cors_preflight_body_is_empty_not_null():
...
+    # WHEN calling the event handler
+    result = handler(mock_event, None)
+
+    # THEN then the response is not compressed
+    assert result["isBase64Encoded"] is False
+    assert result["body"] == expected_value
+    assert result["multiValueHeaders"].get("Content-Encoding") is None
+
+
+def test_compress_response_object():
```
Comment (heitorlessa) [inline tests/functional/event_handler/test_api_gateway.py]:
```suggestion
def test_response_with_compress_enabled():
```

File: aws_lambda_powertools/event_handler/api_gateway.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/api_gateway.py]: I completely agree!! The code is much cleaner now :)

File: tests/functional/event_handler/test_api_gateway.py (hunk 2 above)
Comment (leandrodamascena) [reply tests/functional/event_handler/test_api_gateway.py]: Fixed

File: aws_lambda_powertools/event_handler/api_gateway.py [hunk 3]
```diff
@@ -235,6 +235,28 @@ def _add_cache_control(self, cache_control: str):
         cache_control = cache_control if self.response.status_code == 200 else "no-cache"
         self.response.headers["Cache-Control"] = cache_control
 
+    def _has_compression_enabled(self, compress: Optional[bool], event: BaseProxyEvent) -> bool:
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py]:
```suggestion
    def _has_compression_enabled(self, compress: bool, event: BaseProxyEvent) -> bool:
```

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**